_posts/2017-04-02-generating-sitemaps-with-apache-spark.markdown

```diff
@@ -0,0 +1,228 @@
+    def sitemappable_priority
+      0.5
+    end
+  end
+end
+```
+
+We also used HAML to generate sitemaps.
+
+```haml
+!!! XML
+%urlset{ xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9" }
+  - documents.each do |document|
+      - type.sitemappable_urls(document, self).each do |url|
+        %url
+          %loc=h url
+```
+
+Of course, there were many more implementation details, but adding a sitemap was easy as we made Mongoid models use `Sitemappable`.
+
+Generating sitemaps took a couple of hours and ran nightly.
+
+### The Dynamic Front-End
+
+By 2015 our front-end was fully decoupled from the monolith that now served as the core API service. The sitemap code still lived there and needed to know about an increasing number of URLs, which seemed totally backwards. We wanted an easier way to maintain and update sitemaps and concluded that by generating sitemaps on the fly we would also see fewer 404 errors during crawling as a side effect. Furthermore, sitemaps increasingly needed information from other places, as we tore the monolith apart into micro-services.
+
+Our solution for dynamic sitemaps queried the API for the number of total items in every collection, then generated a sitemap index with an entry for each page of results. A subsequent query from Google for a given page would query the API for the actual data that was needed to produce the requested sitemap. You can see the entire code [in our open-source artsy.net project](https://github.com/artsy/force/tree/2059f8f5e7faa8213221480c1c9cd3c62e9d5ffb/desktop/apps/sitemaps).
+
+This worked for a while, but introduced new challenges. The first was that we needed individual sitemaps to contain thousands of items or have tens of thousands of sitemaps. That meant tens of megabytes of data returned from either multiple API calls or from a very slow single API query. That problem was even more serious for models that required joining data, and was solved by our first attempt at an orchestration layer that would crawl our own API on a schedule. Today we use GraphQL for that purpose, but we never deprecated the system we put in place during this phase of the sitemap rewrite. That system was barely maintained and riddled with its own issues, starting by the fact that the data crawl began timing out as the amount of data increased sharply.
+
+### The Big(gish) Data Solution
+
+In 2016 we started using Apache Spark for calculating similarity graphs and artwork recommendations. We loaded a vast majority of our data from many data sources to HDFS and wrote Spark jobs in Scala. Having all the data from everywhere made our Spark cluster a perfect candidate for a smallish job like sitemaps.
+
+#### A Spark Job
+
+A spark job is a Scala class with an entry point that receives some configuration. A typical sitemap fetches data, partitions it, and generates XML sitemap files. This is automatically parallelized and distributed across worker nodes, making the whole thing blazingly fast.
```

Comment:
You could add `...across worker nodes that already have their subset of data locally.`